### PR TITLE
Replicated: For COPY_GET, only src has data/omap digest, the dst can do.

### PR DESCRIPTION
**Pull request by majianpeng**
Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

### CODE REVIEW

**Comment by majianpeng** [comment]
@liewegas . Have you time review this! Thanks!

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -6507,14 +6507,16 @@ void ReplicatedPG::process_copy_chunk(hobject_t oid, ceph_tid_t tid, int r)
...
+  if (cop->results.is_data_digest() || cop->results.is_omap_digest()) {
+    dout(20) << __func__ << std::hex
+      << " got digest: rx data 0x" << cop->results.data_digest
+      << " omap 0x" << cop->results.omap_digest
+      << ", source: data 0x" << cop->results.source_data_digest
+      << " omap 0x" <<  cop->results.source_omap_digest
+      << std::dec
+      << " flags " << cop->results.flags
+      << dendl;
+  }
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: this makes sense...

File: src/osd/ReplicatedPG.cc [hunk 2]
```diff
@@ -6635,7 +6639,8 @@ void ReplicatedPG::_write_copy_chunk(CopyOpRef cop, PGBackend::PGTransaction *t)
 	// don't checksum the key count prefix
 	bufferlist keys;
 	keys.substr_of(cop->omap_data, 4, cop->omap_data.length() - 4);
-	cop->results.omap_digest = keys.crc32c(cop->results.omap_digest);
+	if (cop->results.is_omap_digest())
+	  cop->results.omap_digest = keys.crc32c(cop->results.omap_digest);
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: i think the rest does not

**Comment by liewegas** [comment]
The is_data_digest() etc helpers tell you whether source_{data,omap}_digest are defined.  but that should't affect whether we calculate a new digest based on the data they send...

**Comment by majianpeng** [comment]
hh. Thanks your suggestion. In the process of reading code, i found some bugs. Can you look those? Thanks!

**Comment by liewegas** [comment]
These all look good to me!  Reviewed-by:

**Comment by athanatos** [comment]
Ping me when you have updated the series.

**Comment by majianpeng** [comment]
@athanatos . updated and please review again! Thanks!

**Comment by athanatos** [comment]
@majianpeng: added to wip-sam-testing again.  It would still be good for you to get access to sepia and learn to tests PRs prior to posting.

**Pull request merged**